### DISCUSSION TITLE
[Q&A] Several questions about Compaction

### DISCUSSION THREAD

**Discussion by SeunghyunYoo** [milvus-io/milvus#28565] [Q&A and General discussion]
I have several questions about compaction.

How `the segment maximum number of rows` is defined, isn't the parameter based on existing collection info?

When i did manual compaction on collection, the situation is followed.

`before compaction`
![image](https://github.com/milvus-io/milvus/assets/124007900/6402aa4e-bee2-4b8a-b29e-728b1333aafa)

`after triggering compaction`
![image](https://github.com/milvus-io/milvus/assets/124007900/c11add02-de11-429a-9c7d-51bdeb3702a5)

`after compaction`
![image](https://github.com/milvus-io/milvus/assets/124007900/d7af9eee-28c1-47c3-8e0c-bfc2bf05688f)

`The segment maximum number of rows` of the collection is `480,000`.

There is no segments which have smaller # of rows than `480,0000 * 0.5 = 240,000`.
According to `milvus config`, all the segment must not to be candidate of compaction plan.

> The segment is considered as "small segment" when its # of rows is smaller than (smallProportion * segment max # of rows).

However, all segments are included in compaction plans, so the segments have to be re-indexed.

```
Compaction Plans:
 - compaction id: 445692609026704070
 - state: Completed
 - plans: [
Plan:
 - sources: [445692608972043003]
 - target: 445692608972043241
, 
Plan:
 - sources: [445692608972043143]
 - target: 445692608972043381
, 
Plan:
 - sources: [445692608972043075]
 - target: 445692608972043313
, 
Plan:
 - sources: [445692608972043002]
 - target: 445692608972043240
, 
Plan:
 - sources: [445692608972043074]
 - target: 445692608972043312
, 
Plan:
 - sources: [445692608972043210]
 - target: 445692608972043448
```

https://github.com/milvus-io/milvus/blob/8b453454dd5691265482adf70b0e3b19ab14a9ce/configs/milvus.yaml#L344-L381

Could you explain the compaction scenario?

My milvus cluster version is `v2.3.2` and i use all defaults configuration.

Plus, i want to know what `indexBasedCompaction` means.

thanks

**Comment by yhmo** [comment]
When you continually insert data into milvus, the data is not persisted into storage instantly. Querynode accumulates the data in memory to perform brute-force search on it. Datanode accumulates the data in a buffer until the buffer size exceeds a threshold. The threshold is dataCoord.segment.maxSize * dataCoord.segment.sealProportion, after the datanode flushes the buffer into storage, the data becomes a "sealed segment". 
The value dataCoord.segment.sealProportion is 0.23, this value should not be too large or too small, the reason is: 
- only "sealed segment" can be indexed, before the "growing segment" is sealed, it always performs brute-force search and slows down the search performance. If this value is large, a large "growing segment" will lead to poor search performance.
- If the value is small, there will be lots of small sealed segments generated, large number of small segments also slows down search performance, and brings heavy workload to datanode to compact them.


The major purposes of compaction:
1. merge small segments into large segments because large amount of small segments slows down search performance. Ideally, the size of each segment is near to dataCoord.segment.maxSize=512MB.
2. in milvus, delete action is "soft-delete". When you call delete(), an entity is marked as "soft-deleted" in a segment, it is invisible/ignored to search/query, but still occupies memory and storage space. With more and more entities are deleted, more memory and storage space is wasted, a compaction can erase the deleted entities and generated a new segment with the non-deleted entities. Then. querynode replaces the old segment by the new segment to save memory.

Some conditions that trigger compaction:
1. if a segment's size is smaller than dataCoord.segment.maxSize * dataCoord.segment.smallProportion, it will be detected as "small segment" and can be compacted.
2. if there are only a few small segments, the sum of their size is smaller than dataCoord.segment.maxSize * dataCoord.segment.compactableProportion, datanode will not trigger compaction.
3. if the number of deleted entities of a segment exceeds a threshold(20%), datanode will trigger a "single-compaction" for this segment.


indexBasedCompaction=true ------------------------ Only compact segments whose index is already finished. Ignore the segments whose index is not finished.
Milvus builds an independent index for each segment. So, if segments are compacted to a new segment, the old segments are marked as "deleted" and waiting for GC, and indexnode will build index for the new segment. Set indexBasedCompaction to true can avoid unnecessary index work.

**Reply by SeunghyunYoo** [reply]
https://github.com/milvus-io/milvus/discussions/28565#discussion-5870776

<img width="1189" alt="image" src="https://github.com/milvus-io/milvus/assets/124007900/08a59903-ddfb-4211-8ef6-c23be36884fd">

The collection is composed of 256 dim float vector, so the minimum segment size of the collection is `351.6MB` (360,000 * 4 * 256 bytes at least)

compaction trigger size = `435.2MB` (`512MB * 0.85`)

plus, there was no upsert/delete operation after building collection first. (i mean all the entities have no delete marker in my situation)

> if a segment's size is smaller than dataCoord.segment.maxSize * dataCoord.segment.smallProportion, it will be detected as "small segment" and can be compacted.

According to your mention, when I trigger manual compaction, compaction must not to be triggered on all the segments.

However, compaction triggered on all the segments in my case.

How can i handle this situation?

**Reply by yhmo** [reply]
There are more conditions that can allow compaction to a segment:
- the stat log files are too many
- the insert-log files are too many

All the conditions are internal logic, I don't think users should understand it.

**Reply by yhmo** [reply]
The major source code to define compaction condition is here: https://github.com/milvus-io/milvus/blob/master/internal/datacoord/compaction_trigger.go

**Reply by SeunghyunYoo** [reply]
> The major source code to define compaction condition is here: https://github.com/milvus-io/milvus/blob/master/internal/datacoord/compaction_trigger.go

thanks :)

**Comment by SeunghyunYoo** [comment]
I follow up compaction step through https://github.com/milvus-io/milvus/blob/master/internal/datacoord/compaction_trigger.go

so close issue
